movement/watch_faces/complication/totp_face.c

```diff
@@ -1,44 +1,180 @@
+    do {
+        if (num_totp_records == MAX_TOTP_RECORDS) {
+            printf("TOTP max records: %d", MAX_TOTP_RECORDS);
+            break;
+        }
+
+        // Check that it looks like a URI
+        if (strncmp(TOTP_URI_START, line, uri_start_len)) {
+            printf("TOTP invalid uri start: %s", line);
+            continue;
+        }
+
+        // Check that we can find a '?' (to start our parameters)
+        char *param;
+        char *param_saveptr = NULL;
+        char *params = strchr(line + uri_start_len, '?');
+        if (params == NULL) {
+            printf("TOTP no params: %s", line);
+            continue;
+        }
+
+        // Process the parameters and put them in the record
+        init_totp_record(&totp_records[num_totp_records]);
+        bool error = false;
+        param = strtok_r(params + 1, "&", &param_saveptr);
+        do {
+            char *param_middle = strchr(param, '=');
+            *param_middle = '\0';
+            error = error || !totp_face_read_param(&totp_records[num_totp_records], param, param_middle + 1);
+        } while ((param = strtok_r(NULL, "&", &param_saveptr)));
+
+        if (error) {
+            if (totp_records[num_totp_records].secret != NULL) {
+                free(totp_records[num_totp_records].secret);
+            }
+            continue;
+        }
+
+        // If we found a probably valid TOTP record, keep it. 
+        if (totp_records[num_totp_records].secret != NULL) {
+            num_totp_records += 1;
+        } else {
+            printf("TOTP missing secret: %s", line);
+        }
+    } while ((line = strtok_r(NULL, "\n", &line_saveptr)));
+
+    free(text);
+}
+
 void totp_face_setup(movement_settings_t *settings, uint8_t watch_face_index, void ** context_ptr) {
     (void) settings;
     (void) watch_face_index;
-    if (*context_ptr == NULL) *context_ptr = malloc(sizeof(totp_state_t));
+    if (*context_ptr == NULL) {
+        *context_ptr = malloc(sizeof(totp_state_t));
+        totp_face_read_file(TOTP_FILE);
+    }
 }
 
 void totp_face_activate(movement_settings_t *settings, void *context) {
     (void) settings;
     memset(context, 0, sizeof(totp_state_t));
     totp_state_t *totp_state = (totp_state_t *)context;
-    TOTP(keys, key_sizes[0], timesteps[0]);
-    totp_state->timestamp = watch_utility_date_time_to_unix_time(watch_rtc_get_date_time(), movement_timezone_offsets[settings->bit.time_zone] * 60);
-    totp_state->current_code = getCodeFromTimestamp(totp_state->timestamp);
+
+    if (num_totp_records > 0) {
+        totp_state->current_index = 0;
+        totp_state->steps = 0;  // Force getting a new code in the loop.
+        totp_state->timestamp = watch_utility_date_time_to_unix_time(watch_rtc_get_date_time(), movement_timezone_offsets[settings->bit.time_zone] * 60);
+        TOTP(totp_records[0].secret, totp_records[0].secret_size, totp_records[0].period);
+    } else {
+        watch_display_string("NA00000000", 0);
```

Comment:
Suggestions welcome for a what a better string is here.

Reply:
I changed this to read "No 2F codes" across both lines.